src/t_zset.c

```diff
@@ -328,54 +329,57 @@ int zslIsInRange(zskiplist *zsl, zrangespec *range) {
     return 1;
 }
 
-/* Find the first node that is contained in the specified range.
+/* Find the Nth node that is contained in the specified range. N should be 0-based.
  * Returns NULL when no element is contained in the range. */
-zskiplistNode *zslFirstInRange(zskiplist *zsl, zrangespec *range) {
+zskiplistNode *zslNthInRange(zskiplist *zsl, zrangespec *range, long n) {
     zskiplistNode *x;
     int i;
 
     /* If everything is out of range, return early. */
     if (!zslIsInRange(zsl,range)) return NULL;
 
     x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *OUT* of range. */
-        while (x->level[i].forward &&
-            !zslValueGteMin(x->level[i].forward->score,range))
+    if (n >= 0) {
+        for (i = zsl->level - 1; i >= 0; i--) {
+            if (!x->level[0].forward || zslValueGteMin(x->level[0].forward->score, range)) {
+                break;
+            }
+            /* Go forward while *OUT* of range. */
+            while (x->level[i].forward && !zslValueGteMin(x->level[i].forward->score, range)) {
                 x = x->level[i].forward;
-    }
-
-    /* This is an inner range, so the next node cannot be NULL. */
-    x = x->level[0].forward;
-    serverAssert(x != NULL);
-
-    /* Check if score <= max. */
-    if (!zslValueLteMax(x->score,range)) return NULL;
-    return x;
-}
-
-/* Find the last node that is contained in the specified range.
- * Returns NULL when no element is contained in the range. */
-zskiplistNode *zslLastInRange(zskiplist *zsl, zrangespec *range) {
-    zskiplistNode *x;
-    int i;
-
-    /* If everything is out of range, return early. */
-    if (!zslIsInRange(zsl,range)) return NULL;
-
-    x = zsl->header;
-    for (i = zsl->level-1; i >= 0; i--) {
-        /* Go forward while *IN* range. */
-        while (x->level[i].forward &&
-            zslValueLteMax(x->level[i].forward->score,range))
+            }
+        }
+        for (i++, n++; i >= 0 && n > 0; i--) {
+            while (x->level[i].forward && (unsigned long)n >= x->level[i].span) {
+                n -= x->level[i].span;
+                x = x->level[i].forward;
+            }
```

Comment:
please add some comments.
i don't understand why you advance to the next level (`i++` on `i` left from the previous loop), or why is it safe.

Reply:
If we left last loop with i==-1, we need a i++ to go back to level[0].
If we left last loop from the break, it might not be save since i can be zsl->level-1...

However I already rewrote this part, please check the new PR.